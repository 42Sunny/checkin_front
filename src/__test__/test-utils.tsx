// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
