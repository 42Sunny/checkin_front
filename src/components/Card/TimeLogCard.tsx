import React from "react";

import dayjs from "dayjs";
import localizedFormat from "dayjs/plugin/localizedFormat";

import AccountBoxIcon from "@mui/icons-material/AccountBox";
import classes from "../../styles/components/Card/TimeLogCard.module.css";

dayjs.extend(localizedFormat);
dayjs.locale("ko");

type UtilBoxProps = {
  handleFlip: () => void;
};

function UtilBox({ handleFlip }: UtilBoxProps) {
  return (
    <div className={classes["util-box"]}>
      <AccountBoxIcon onClick={handleFlip} />
    </div>
  );
}

const data = {
  durationTime: 234000,
  logs: [
    {
      id: 1,
      checkinAt: "2020-07-10T13:00:00",
      checkoutAt: "2020-07-10T18:00:00",
      durationTime: 3600,
    },
    {
      id: 2,
      checkinAt: "2020-07-11T13:00:00",
      checkoutAt: "2020-07-11T18:00:00",
      durationTime: 10860,
    },
  ],
};

const timeFormat = (seconds: number) => {
  const tempHours = Math.floor(seconds / 3600);
  const temppMinuts = (seconds % 3600) / 60;

  const hours = tempHours < 10 ? `0${tempHours}` : tempHours;
  const minutes = temppMinuts < 10 ? `0${temppMinuts}` : temppMinuts;
  return `${hours}시간${minutes}분`;
};

function LogCardContents() {
  const accTime = data.logs.reduce((ac, cur) => ac + cur.durationTime, 0);
  return (
    <div className={classes.timeLogWrpaper}>
      <div className={classes.timeLogTitleWrapper}>
        <p>총 누적 시간 : {timeFormat(accTime)}</p>
        <strong>이번달</strong>
      </div>
      <table className={classes.timeLogTable}>
        <thead>
          <tr>
            <th>입실</th>
            <th>퇴실</th>
            <th>체류시간</th>
          </tr>
        </thead>
        <tbody>
          {data.logs.map((log: any) => {
            return (
              <tr key={log.id} className={classes.row}>
                <td>{dayjs(log.checkinAt).format("MM월DD일 HH:mm")}</td>
                <td>{dayjs(log.checkoutAt).format("MM월DD일 HH:mm")}</td>
                <td>{timeFormat(log.durationTime)}</td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
}

type TimeLogCardProps = {
  handleFlip: () => void;
  logs: Log[];
};

function TimeLogCard({ handleFlip, logs }: TimeLogCardProps) {
  return (
    <div className={classes["time-log-wrapper"]}>
      <UtilBox handleFlip={handleFlip} />
      <LogCardContents />
    </div>
  );
}

export default TimeLogCard;
