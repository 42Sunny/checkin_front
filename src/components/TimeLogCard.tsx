import AccountBoxIcon from "@mui/icons-material/AccountBox";
import moment from "moment-timezone";
import React, { useEffect, useState } from "react";
import classes from "../styles/components/TimeLogCard.module.css";
import useUser from "../utils/hooks/useUser";

interface HeaderProps {
  checkinTime: string;
}
const Header: React.FC<HeaderProps> = ({ checkinTime }) => (
  <div className={classes["time-log-header"]}>
    <h4 className={classes["time-log-title"]}>클러스터 로그</h4>
    <p className={classes["time-log-subtitle"]}>체크인 시각: {checkinTime}</p>
  </div>
);
interface UtilBoxProps {
  handleFlip: () => void;
}
const UtilBox: React.FC<UtilBoxProps> = ({ handleFlip }) => (
  <div className={classes["util-box"]}>
    <AccountBoxIcon onClick={handleFlip} />
  </div>
);

interface LogDataProps {
  log: Log;
}

const LogData: React.FC<LogDataProps> = ({ log: { date, seconds } }) => (
  <li className={classes["log-data"]}>
    <time dateTime={date}>{date}</time>
    <div>{moment.utc(+seconds * 1000).format("HH:mm:ss")}</div>
  </li>
);

interface LogDataListProps {
  logs: Log[];
}

const LogDataList: React.FC<LogDataListProps> = ({ logs }) => (
  <>
    <li className={classes["log-data-header"]}>
      <div>날짜</div>
      <div>이용 시간</div>
    </li>
    <hr className={classes.divider} />
    <ul className={classes["log-data-wrapper"]}>
      {logs.map((log, idx) => (
        <div key={idx.toString()}>
          <LogData log={log} />
          <hr className={classes.divider} />
        </div>
      ))}
    </ul>
  </>
);
interface IProps {
  handleFlip: () => void;
  logs: Log[];
}

const TimeLogCard: React.FC<IProps> = ({ handleFlip, logs }) => {
  const {
    user: { checkinAt },
  } = useUser();
  const [checkinTime, setCheckinTime] = useState("");

  useEffect(() => {
    if (checkinAt) setCheckinTime(moment(new Date(checkinAt)).format("YYYY-MM-DD HH:mm"));
  }, [checkinAt]);

  return (
    <div className={classes["time-log-wrapper"]}>
      <Header checkinTime={checkinTime} />
      <UtilBox handleFlip={handleFlip} />
      <LogDataList logs={logs} />
    </div>
  );
};

export default TimeLogCard;
