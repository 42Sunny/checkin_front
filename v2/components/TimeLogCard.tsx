import AccountBoxIcon from "@mui/icons-material/AccountBox";
import moment from "moment-timezone";
import React, { useEffect, useState } from "react";
import classes from "../styles/components/TimeLogCard.module.css";
import useUser from "../utils/hooks/useUser";
import { formatToGeneralTime } from "../utils/time";

interface HeaderProps {
  checkinTime: string;
}
const Header: React.FC<HeaderProps> = ({ checkinTime }) => (
  <div className={classes["time-log-header"]}>
    <h4 className={classes["time-log-title"]}>클러스터 로그</h4>
    <p className={classes["time-log-subtitle"]}>체크인 시각: {checkinTime}</p>
  </div>
);
interface UtilBoxProps {
  handleFlip: () => void;
}
const UtilBox: React.FC<UtilBoxProps> = ({ handleFlip }) => (
  <div className={classes["util-box"]}>
    <AccountBoxIcon onClick={handleFlip} />
  </div>
);

interface LogDataProps {
  log: Log;
}

const LogData: React.FC<LogDataProps> = ({ log: { created_at, duration } }) => {
  const date = created_at.split("T")[0];
  return (
    <li className={classes["log-data"]}>
      <time dateTime={date}>{date}</time>
      <div>{moment.utc(duration * 1000).format("HH:mm:ss")}</div>
    </li>
  );
};

interface LogDataListProps {
  logs: Log[];
}

const LogDataList: React.FC<LogDataListProps> = ({ logs }) => {
  return (
    <>
      <li className={classes["log-data-header"]}>
        <div>날짜</div>
        <div>이용 시간</div>
      </li>
      <hr className={classes.divider} />
      <ul className={classes["log-data-wrapper"]}>
        {logs.map((log, idx) => (
          <div key={idx.toString()}>
            <LogData log={log} />
            <hr className={classes.divider} />
          </div>
        ))}
      </ul>
    </>
  );
};
interface IProps {
  handleFlip: () => void;
  logs: Log[];
}

const TimeLogCard: React.FC<IProps> = ({ handleFlip, logs }) => {
  const {
    user: { checkinAt },
  } = useUser();
  const [checkinTime, setCheckinTime] = useState("");

  useEffect(() => {
    if (checkinAt) setCheckinTime(formatToGeneralTime(new Date(checkinAt)));
  }, [checkinAt]);
  return (
    <div className={classes["time-log-wrapper"]}>
      <Header checkinTime={checkinTime} />
      <UtilBox handleFlip={handleFlip} />
      <LogDataList logs={logs} />
    </div>
  );
};

export default TimeLogCard;
