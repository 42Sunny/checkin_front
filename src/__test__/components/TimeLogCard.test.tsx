import TimeLogCard from "../../components/TimeLogCard";
import { screen, createWithWrapper, render } from "../test-utils";

describe("src/components/TimeLogCard.tsx", () => {
  const timeLogCardProps: { handleFlip: any; logs: any } = { handleFlip: "", logs: "" };
  beforeEach(() => {
    timeLogCardProps.handleFlip = jest.fn();
    timeLogCardProps.logs = [];
  });
  it("matches snapshot", () => {
    const component = createWithWrapper(<TimeLogCard {...timeLogCardProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
