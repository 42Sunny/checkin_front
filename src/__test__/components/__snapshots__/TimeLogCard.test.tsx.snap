// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/TimeLogCard.tsx matches snapshot 1`] = `
<div
  className="time-log-wrapper"
>
  <div
    className="time-log-header"
  >
    <h4
      className="time-log-title"
    >
      클러스터 로그
    </h4>
    <p
      className="time-log-subtitle"
    >
      체크인 시각: 
      
    </p>
  </div>
  <div
    className="util-box"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="AccountBoxIcon"
      focusable="false"
      onClick={[MockFunction]}
      viewBox="0 0 24 24"
    >
      <path
        d="M3 5v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2H5c-1.11 0-2 .9-2 2zm12 4c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3zm-9 8c0-2 4-3.1 6-3.1s6 1.1 6 3.1v1H6v-1z"
      />
    </svg>
  </div>
  <li
    className="log-data-header"
  >
    <div>
      날짜
    </div>
    <div>
      이용 시간
    </div>
  </li>
  <hr
    className="divider"
  />
  <ul
    className="log-data-wrapper"
  />
</div>
`;
